File: jinq-jpa/test/org/jinq/jpa/JinqJPATestBase.java
package org.jinq.jpa;

import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.Persistence;

import org.jinq.jpa.test.entities.Lineorder;
import org.jinq.jpa.test.entities.PhoneNumber;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;

public class JinqJPATestBase
{
   // TODO: If these are static, then multiple tests can't be run
   //   simultaneously.
   static EntityManagerFactory entityManagerFactory;
   static JinqJPAStreamProvider streams;

   EntityManager em;
   String query;
   List<String> queryList = new ArrayList<>();
   
   
   @BeforeClass
   public static void setUpBeforeClass() throws Exception
   {
      entityManagerFactory = Persistence.createEntityManagerFactory("JPATest");
      streams = new JinqJPAStreamProvider(entityManagerFactory);
      
      // Hibernate seems to generate incorrect metamodel data for some types of
      // associations, so we have to manually supply the correct information here.
      streams.registerAssociationAttribute(Lineorder.class.getMethod("getItem"), "item", false);
      streams.registerAssociationAttribute(Lineorder.class.getMethod("getSale"), "sale", false);
      
      // Register types that are used by AttributeConverters
      streams.registerAttributeConverterType(PhoneNumber.class);
      
      EntityManager em = entityManagerFactory.createEntityManager();
      new CreateJpaDb(em).createDatabase();
      em.close();
   }

   @AfterClass
   public static void tearDownAfterClass() throws Exception
   {
      entityManagerFactory.close();
      try {
         DriverManager.getConnection("jdbc:derby:memory:demoDB;drop=true");
      } catch (SQLException e) { }
   }
   
   @Before
   public void setUp() throws Exception
   {
      em = entityManagerFactory.createEntityManager();
      em.getTransaction().begin();
      queryList.clear();
      streams.setHint("exceptionOnTranslationFail", true);
      streams.setHint("queryLogger", new JPAQueryLogger() {
         @Override public void logQuery(String q,
               Map<Integer, Object> positionParameters,
               Map<String, Object> namedParameters)
         {
            queryList.add(q);
            query = q;
         }});
   }

   @After
   public void tearDown() throws Exception
   {
      em.getTransaction().rollback();
      em.close();
   }
}

File: jinq-jpa-scala/src/test/scala/org/jinq/jpa/JinqJPATestBase.scala
package org.jinq.jpa;

import java.sql.DriverManager
import java.sql.SQLException
import java.util.Map
import _root_.scala.collection.mutable.ArrayBuffer
import org.jinq.jpa.test.entities.Lineorder
import org.junit.After
import org.junit.AfterClass
import org.junit.Before
import org.junit.BeforeClass
import javax.persistence.EntityManager
import javax.persistence.EntityManagerFactory
import javax.persistence.Persistence
import org.jinq.jpa.test.entities.PhoneNumber

object JinqJPAScalaTestBase {
  var entityManagerFactory: EntityManagerFactory = _
  var streams : JinqJPAScalaIteratorProvider = _

   @BeforeClass
   def setUpBeforeClass() {
      entityManagerFactory = Persistence.createEntityManagerFactory("JPATest");
      streams = new JinqJPAScalaIteratorProvider(JinqJPAScalaTestBase.entityManagerFactory);
      
      // Hibernate seems to generate incorrect metamodel data for some types of
      // associations, so we have to manually supply the correct information here.
      streams.registerAssociationAttribute(classOf[Lineorder].getMethod("getItem"), "item", false);
      streams.registerAssociationAttribute(classOf[Lineorder].getMethod("getSale"), "sale", false);
      
      // Register types that are used by AttributeConverters
      streams.registerAttributeConverterType(classOf[PhoneNumber]);

      var em = entityManagerFactory.createEntityManager();
      new CreateJpaDb(em).createDatabase();
      em.close();
   }

   @AfterClass
   def tearDownAfterClass() {
      entityManagerFactory.close();
      try {
         DriverManager.getConnection("jdbc:derby:memory:demoDB;drop=true");
      } catch {
        case e: SQLException => { }
      }
   }
   

}
class JinqJPAScalaTestBase {
   var em : EntityManager = null;
   var query : String = null;
   val queryList = new ArrayBuffer[String]();
   
   @Before
   def setUp() {
      em = JinqJPAScalaTestBase.entityManagerFactory.createEntityManager();
      em.getTransaction().begin();
      queryList.clear();
      query = null
      JinqJPAScalaTestBase.streams.setHint("exceptionOnTranslationFail", java.lang.Boolean.TRUE);
      JinqJPAScalaTestBase.streams.setHint("queryLogger", new JPAQueryLogger() {
         def logQuery(q: String,
               positionParameters: Map[Integer, Object] ,
               namedParameters: Map[String, Object] )
         {
            queryList.append(q);
            query = q;
         }});
   }

   @After
   def tearDown() {
      em.getTransaction().rollback();
      em.close();
   }
}
